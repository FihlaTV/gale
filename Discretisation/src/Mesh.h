/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003-2006, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street,
**	Melbourne, 3053, Australia.
**
** Primary Contributing Organisations:
**	Victorian Partnership for Advanced Computing Ltd, Computational Software Development - http://csd.vpac.org
**	Australian Computational Earth Systems Simulator - http://www.access.edu.au
**	Monash Cluster Computing - http://www.mcc.monash.edu.au
**	Computational Infrastructure for Geodynamics - http://www.geodynamics.org
**
** Contributors:
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Robert Turnbull, Research Assistant, Monash University. (robert.turnbull@sci.monash.edu.au)
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	David May, PhD Student, Monash University (david.may@sci.monash.edu.au)
**	Louis Moresi, Associate Professor, Monash University. (louis.moresi@sci.monash.edu.au)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**	Julian Giordani, Research Assistant, Monash University. (julian.giordani@sci.monash.edu.au)
**	Vincent Lemiale, Postdoctoral Fellow, Monash University. (vincent.lemiale@sci.monash.edu.au)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
*/
/** \file
**  Role:
**	Brings together the geometry and node interconnectivity of a mesh, plus the element type (i.e. shape functions) needed by the finite element method.
**
** Assumptions:
**	Elements used with this mesh inherit from FiniteElement_Element.
**	The contents of the mesh's elements must be initialised by the user.
**
** Comments:
**	This mesh inherits from the parent Discretisation-level Mesh class, which defines how
**	the mesh is laid out, manages local to global mappings etc.
**	The extra complexity this class adds from an FE perspective is the relationship between
**	the geometric element /node types to the F.E. ElementType determining shape functions.
**
**	Situations involving this class become interesting when using multiple FeVariable objects
**	in a simulation. See the Twiki notes at 
**	csd.vpac.org/twiki/bin/view/Stgermain/FiniteElement_Mesh for more on this.
**
** $Id $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#ifndef __StgFEM_Discretisation_FiniteElement_Mesh_h__
#define __StgFEM_Discretisation_FiniteElement_Mesh_h__
	
	/** Textual name of this class */
	extern const Type FiniteElement_Mesh_Type;
	
	/** FiniteElement_Mesh information */
	#define __FiniteElement_Mesh \
		/* General info */ \
		__Mesh \
		\
		/* Virtual info */ \
		\
		/* FiniteElement_Mesh info */ \
		/** The basic discretisation Mesh */ \
		ElementType_Register*			elementType_Register; \
		
	/** Brings together and manages the life-cycle of a a mesh and all the 
	info relevant to it for the Finite Element Method. See Mesh.h for more. */
	struct FiniteElement_Mesh { __FiniteElement_Mesh };
	
	/* --- Constructors / Destructor --- */
	
	/** Create a new FiniteElement_Mesh and initialise */

	void* FiniteElement_Mesh_DefaultNew( Name name );
	
	FiniteElement_Mesh* FiniteElement_Mesh_New(
		Name					name,
		void*					layout,
		SizeT					nodeSize,
		SizeT					elementSize,
		void*					extension_Register,
		void*					elementType_Register,
		Dictionary*				dictionary );

	void FiniteElement_Mesh_LoadFromDict( void* mesh, Dictionary* subDict, Dictionary* dictionary, Stg_ObjectList* objList);
	
	/* Initialise a FiniteElement_Mesh construct */
	void FiniteElement_Mesh_Init(
		FiniteElement_Mesh*			self,
		Name					name,
		void*					layout,
		SizeT					nodeSize,
		SizeT					elementSize,
		void*					extension_Register,
		void*					elementType_Register,
		Dictionary*				dictionary );
	
	/* Creation implementation / Virtual constructor */
	FiniteElement_Mesh* _FiniteElement_Mesh_New(
		SizeT					_sizeOfSelf,
		Type					type,
		Stg_Class_DeleteFunction*			_delete,
		Stg_Class_PrintFunction*			_print,
		Stg_Class_CopyFunction*			_copy,
		Stg_Component_DefaultConstructorFunction*	_defaultConstructor,
		Stg_Component_ConstructFunction*	_construct,
		Stg_Component_BuildFunction*		_build,
		Stg_Component_InitialiseFunction*		_initialise,
		Stg_Component_ExecuteFunction*		_execute,
		Stg_Component_DestroyFunction*		_destroy,
		Name					name,
		Bool					initFlag,
		Mesh_Node_IsLocalFunction*		_nodeIsLocal,
		Mesh_Node_IsShadowFunction*		_nodeIsShadow,
		Mesh_Element_IsLocalFunction*		_elementIsLocal,
		Mesh_Element_IsShadowFunction*		_elementIsShadow,
		void*					layout,
		SizeT					nodeSize,
		SizeT					elementSize, 
		void*					extension_Register,
		void*					elementType_Register,
		Dictionary*				dictionary );

	/* Initialise implementation */
	void _FiniteElement_Mesh_Init( 
		FiniteElement_Mesh*			self,
		void*					elementType_Register );
	
	/* Stg_Class_Delete a FiniteElement_Mesh construst */
	void _FiniteElement_Mesh_Delete( void* feMesh );

	/* --- Virtual Function implementations --- */
	
	/* Print the contents of an FiniteElement_Mesh construct */
	void _FiniteElement_Mesh_Print( void* feMesh, Stream* stream );
	
	/* Copy */
	#define FiniteElement_Mesh_Copy( self ) \
		(FiniteElement_Mesh*)Stg_Class_Copy( self, NULL, False, NULL, NULL )
	#define FiniteElement_Mesh_DeepCopy( self ) \
		(FiniteElement_Mesh*)Stg_Class_Copy( self, NULL, True, NULL, NULL )
	
	void* _FiniteElement_Mesh_Copy( void* feMesh, void* dest, Bool deep, Name nameExt, PtrMap* ptrMap );
	
	/* Build implementation */
	void _FiniteElement_Mesh_Build( void* feMesh, void* data );
	
	/* Construct implementation */
	void _FiniteElement_Mesh_Construct( void* feMesh, Stg_ComponentFactory* cf, void* data );
	
	/* Initialisation implementation */
	void _FiniteElement_Mesh_Initialise( void* feMesh, void* data );
	
	/* Execution implementation */
	void _FiniteElement_Mesh_Execute( void* feMesh, void* data );
	
	/* Destruct Implementation */
	void _FiniteElement_Mesh_Destroy( void* feMesh, void* data );

	/* --- Public Functions --- */
	
	/* For use when Node is not yet a complete type */
	#define FiniteElement_Mesh_CoordAt( self, node_dI ) \
		Mesh_CoordAt( self, node_dI )
	double* _FiniteElement_Mesh_CoordAt( void* feMesh, Node_DomainIndex node_dI );
	
	/* For use when Node is not yet a complete type */
	#define FiniteElement_Mesh_NodeAt( self, node_dI ) \
		Mesh_NodeAt( self, node_dI )
	Node* _FiniteElement_Mesh_NodeAt( void* feMesh, Node_DomainIndex node_I );
	
	/* For use when Element is not yet a complete type */
	#define FiniteElement_Mesh_ElementAt( self, element_I ) \
		((FiniteElement_Element*)Mesh_ElementAt( self, element_I ))
	FiniteElement_Element* _FiniteElement_Mesh_ElementAt( void* feMesh, Element_DomainIndex element_I );
	
	/** Get the element type of an element */
	#define FiniteElement_Mesh_ElementTypeAt( self, el_I ) \
		(ElementType_Register_At( (self)->elementType_Register, FiniteElement_Mesh_ElementAt( self, el_I )->elementType_I ))
	ElementType* _FiniteElement_Mesh_ElementTypeAt( void* feMesh, Element_DomainIndex element_I );
	
	/* --- Private Functions --- */
	
	void _FiniteElement_Mesh_SetupElementTypes_Default( FiniteElement_Mesh* self );

	/* --- Public Functions --- */
	Variable* FiniteElement_Mesh_RegisterNodeCoordsAsVariables( void* feMesh, void* _variable_Register, Variable** variableList ) ;
	Variable* FiniteElement_Mesh_RegisterLocalNodeCoordsAsVariables( void* feMesh, void* _variable_Register, Variable** variableList );
	void FiniteElement_Mesh_CalcGlobalCoordFromLocalCoord( void* feMesh, Dimension_Index dim, Element_LocalIndex lElement_I, Coord xi, Coord globalCoord ) ;

	/** This is a higher-level function to the actual one on the elementType that does the work.
		It was decided the user should pass in the current element, since they usually know this when doing this
		conversion */
	void FiniteElement_Mesh_CalcLocalCoordFromGlobalCoord(
		void* feMesh,
		Element_DomainIndex elementCoordIn,
		const Coord globalCoord,
		Coord elLocalCoord );

	
#endif /* __StgFEM_Discretisation_FiniteElement_Mesh_h__ */
