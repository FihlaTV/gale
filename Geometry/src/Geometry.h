/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
*/
/** \file
**  Role:
**	External header file to this library.
**
** Assumptions:
**	None so far.
**
** Comments:
**	None so far.
**
** $Id: Geometry.h 4081 2007-04-27 06:20:07Z LukeHodkinson $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#ifndef __Domain_Geometry_h__
#define __Domain_Geometry_h__
	
	#include "units.h"
	#include "types.h"
	#include "isinf.h"
	#include "Dimension.h"
	#include "VectorMath.h"
	#include "TensorMath.h"
	#include "TrigMath.h"
	#include "ComplexMath.h"
	#include "Plane.h"
	#include "Edge.h"
	#include "Line.h"
	#include "RMatrix.h"
	#include "Topology.h"
	#include "IJKTopology.h"
	#include "IJK6Topology.h"
	#include "IJK26Topology.h"
	#include "IrregTopology.h"
	#include "GeometryClass.h"
	#include "BlockGeometry.h"
	#include "RefinedRegionsGeometry.h"
	#include "ShellGeometry.h"
	#include "IrregGeometry.h"
	#include "QuadEdge.h"
	#include "Delaunay.h"
	#include "ParallelDelaunay.h"
	#include "ComplexVectorMath.h"
	#include "stg_lapack.h"
	#include "FullTensorMath.h"
	#include "TensorMultMath.h"
	#include "Simplex.h"
	#include "Hex.h"
	#include "Init.h"
	#include "Finalise.h"

#endif /* __Domain_Geometry_h__ */
