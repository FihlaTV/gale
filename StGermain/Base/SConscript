import os
Import('env')

# Execute subdirectory SConscripts.
objs = SConscript('Foundation/SConscript', exports='env')
objs += SConscript('IO/SConscript', exports='env')
objs += SConscript('Container/SConscript', exports='env')
objs += SConscript('Automation/SConscript', exports='env')
objs += SConscript('Extensibility/SConscript', exports='env')
objs += SConscript('Context/SConscript', exports='env')
objs += SConscript('src/SConscript', exports='env')

# Build libraries.
libs = env.Library(env['helpers']['buildLibDir'](env), objs)
libs += env.LoadableModule(env['helpers']['buildLibDir'](env), objs)
env.Install(os.path.join(env['prefix'], 'lib'), libs)

# Build tests.
SConscript('tests/SConscript', exports='env')

# Build programs.
SConscript('FlattenXML/SConscript', exports='env')

Return('objs')
