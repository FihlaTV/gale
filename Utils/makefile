# obtain defaults for required variables according to system and project location, and then run the build.
ifndef PROJ_ROOT
	PROJ_ROOT=..
endif
include ${PROJ_ROOT}/Makefile.system

subdirs = src tests

include ${PROJ_ROOT}/Makefile.vmake
