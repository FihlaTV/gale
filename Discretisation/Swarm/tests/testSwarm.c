/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** Role:
**	Test that the ElementCellLayout has the same layout and geometry as the mesh's element layout.
**
** Assumptions:
**	None as yet.
**
** Comments:
**	None as yet.
**
** $Id: testSwarm.c 4118 2007-05-22 02:01:25Z RaquibulHassan $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include "Base/Base.h"

#include "Discretisation/Geometry/Geometry.h"
#include "Discretisation/Shape/Shape.h"
#include "Discretisation/Mesh/Mesh.h"
#include "Discretisation/Utils/Utils.h"
#include "Discretisation/Swarm/Swarm.h"

#include <stdio.h>
#include <stdlib.h>


struct _Node {
	Coord				coord;
};

struct _Element {
	Coord				coord;
};

struct _Particle {
	__GlobalParticle
};

Mesh* buildMesh( unsigned nDims, unsigned* size, 
		     double* minCrds, double* maxCrds, 
		     ExtensionManager_Register* emReg )
{
	CartesianGenerator*	gen;
	Mesh*			mesh;

	gen = CartesianGenerator_New( "" );
	CartesianGenerator_SetDimSize( gen, nDims );
	CartesianGenerator_SetTopologyParams( gen, size, 0, NULL, NULL );
	CartesianGenerator_SetGeometryParams( gen, minCrds, maxCrds );

	mesh = Mesh_New( "" );
	Mesh_SetExtensionManagerRegister( mesh, emReg );
	Mesh_SetGenerator( mesh, gen );

	Build( mesh, NULL, False );
	Initialise( mesh, NULL, False );

	KillObject( mesh->generator );

	return mesh;
}

void TestParticleSearchFunc( Swarm* swarm, Coord coord, Stream* stream ) {
	double              distance;
	Particle_Index      closestParticle_I;
	Particle_Index      lParticle_I;
	GlobalParticle*     particle;

	Journal_Printf( stream, "Testing coord %g %g %g\n", coord[ I_AXIS ], coord[ J_AXIS ], coord[ K_AXIS ] );

	closestParticle_I = Swarm_FindClosestParticle( swarm, 3, coord, &distance );

	if ( closestParticle_I < swarm->particleLocalCount ) {
		Journal_Printf( stream, "Closest Particle is %u with distance %g\n", closestParticle_I, distance );

		for ( lParticle_I = 0 ; lParticle_I < swarm->particleLocalCount ; lParticle_I++ ) {
			/* don't bother with testing same particle */
			if (lParticle_I == closestParticle_I)
				continue;

			particle = (GlobalParticle*)Swarm_ParticleAt( swarm , lParticle_I );

			/* Test if particle is closer */
			if (distance > StGermain_DistanceBetweenPoints( coord, particle->coord, 3 ) )
				Journal_Printf( stream, "Particle %u is closer to coord and has distance %g.\n" , 
						lParticle_I, StGermain_DistanceBetweenPoints( coord, particle->coord, 3 ) );
		}
	}
	else
		Journal_Printf( stream, "Coord not found on this processor.\n" );
}

int main( int argc, char* argv[] ) {
	MPI_Comm			CommWorld;
	int				rank;
	int				numProcessors;
	int				procToWatch;
	Dictionary*			dictionary;
	unsigned	nDims = 3;
	unsigned	meshSize[3] = {2, 3, 2};
	double		minCrds[3] = {0.0, 0.0, 0.0};
	double		maxCrds[3] = {300.0, 12.0, 300.0};
	ExtensionManager_Register*		extensionMgr_Register;
	Mesh*				mesh;
	ElementCellLayout*		elementCellLayout;
	RandomParticleLayout*		randomParticleLayout;
	Swarm*				swarm;
	Stream*				stream;
	Coord                coord = { 120, 2, 210 };

	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_dup( MPI_COMM_WORLD, &CommWorld );
	MPI_Comm_size( CommWorld, &numProcessors );
	MPI_Comm_rank( CommWorld, &rank );
	
	Base_Init( &argc, &argv );
	
	DiscretisationGeometry_Init( &argc, &argv );
	DiscretisationShape_Init( &argc, &argv );
	DiscretisationMesh_Init( &argc, &argv );
	DiscretisationUtils_Init( &argc, &argv );
	DiscretisationSwarm_Init( &argc, &argv );
	MPI_Barrier( CommWorld ); /* Ensures copyright info always come first in output */

	stream = Journal_Register (Info_Type, "myStream");

	if( argc >= 2 ) {
		procToWatch = atoi( argv[1] );
	}
	else {
		procToWatch = 0;
	}
	if( rank == procToWatch ) printf( "Watching rank: %i\n", rank );
	
	/* Read input */
	dictionary = Dictionary_New();
	
	/* Init mesh */
	extensionMgr_Register = ExtensionManager_Register_New();
	mesh = buildMesh( nDims, meshSize, minCrds, maxCrds, extensionMgr_Register );
	
	/* Build the mesh */
	Build( mesh, 0, False );
	Initialise( mesh, 0, False );
	
	/* Configure the element-cell-layout */
	elementCellLayout = ElementCellLayout_New( "elementCellLayout", mesh );
	
	/* Configure the random-particle-layout */
	randomParticleLayout = RandomParticleLayout_New( "randomParticleCellLayout", 4, 13 );
	
	/* Configure the swarm */
	swarm = Swarm_New( "testSwarm", elementCellLayout, randomParticleLayout, 3, sizeof(Particle),
		extensionMgr_Register, NULL, CommWorld );
	
	/* Build the swarm */
	Build( swarm, 0, False );
	Initialise( swarm, 0, False );
	
	if( rank == procToWatch ) {
		Stg_Class_Print( swarm, stream );

		Journal_Printf( stream, "\n" );
		Swarm_PrintParticleCoords( swarm, stream );
		Journal_Printf( stream, "\n" );
		Swarm_PrintParticleCoords_ByCell( swarm, stream );
	
		/* Test Particle Searching Function */
		if (numProcessors == 1) {
			coord[ I_AXIS ] = 0.6 * 300; coord[ J_AXIS ] = 0.2*12; coord[ K_AXIS ] = 0.820*300;
			TestParticleSearchFunc( swarm, coord, stream );

			/* Test New Coord */
			coord[ I_AXIS ] = 0.2*300; coord[ J_AXIS ] = 0.9*12; coord[ K_AXIS ] = 0.120*300;
			TestParticleSearchFunc( swarm, coord, stream );
		}
	}
	
	/* Destroy stuff */
	Stg_Class_Delete( swarm );
	Stg_Class_Delete( randomParticleLayout );
	Stg_Class_Delete( elementCellLayout );
	Stg_Class_Delete( mesh );
	Stg_Class_Delete( extensionMgr_Register );
	Stg_Class_Delete( dictionary );
	
	DiscretisationSwarm_Finalise();
	DiscretisationUtils_Finalise();
	DiscretisationMesh_Finalise();
	DiscretisationShape_Finalise();
	DiscretisationGeometry_Finalise();
	
	Base_Finalise();
	
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}
