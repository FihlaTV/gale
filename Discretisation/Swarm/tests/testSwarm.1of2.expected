StGermain Framework revision 4125. Copyright (C) 2003-2005 VPAC.
Cartesian generator: ''
	Target mesh: ''
	Global element size: 2x3x2
	Local offset of rank 0: 0x0x0
	Local range of rank 0: 2x2x2
	Generating vertices...
		... done.
	Generating elements...
		... done.
	Generating element-vertex incidence...
		... done.
	Generating vertex neighbours...
		... done.
	Generating geometry...
		... done.
	Generating element types...
		... element types are 'Mesh_HexType',
		... mesh algorithm type is 'Mesh_RegularAlgorithms',
		... done.
Watching rank: 1
Swarm (ptr): 0x807da10
Stg_Component (ptr): 0x807da10
Stg_Object (ptr): 0x807da10
	Stg_Class (ptr): 0x807da10
		sizeOfSelf: 244
		_deleteSelf: Yes
		type: Swarm
		_delete (func ptr): 0xb7f3f5a6
		_print (func ptr): 0xb7f3f783
		_copy (func ptr): 0xb7f3fd6e
	name: testSwarm
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7f409f1
	_construct (func ptr): 0xb7f40ad0
	_build (func ptr): 0xb7f40e24
	_initialise (func ptr): 0xb7f40f34
	_execute (func ptr): 0xb7f40fc3
	_destroy (func ptr): 0xb7f40fc8
	isConstructed: True
	isBuilt: True
	isInitialised: True
	hasExecuted: False
	isDestroyed: False
	constructor function name: testSwarm-Construct
	build function name: testSwarm-Build
	initialise function name: testSwarm-Initialise
	execute function name: testSwarm-Execute
	destroy function name: testSwarm-Destroy
ElementCellLayout (ptr): 0x8072018
CellLayout (ptr): 0x8072018
Stg_Component (ptr): 0x8072018
Stg_Object (ptr): 0x8072018
	Stg_Class (ptr): 0x8072018
		sizeOfSelf: 156
		_deleteSelf: Yes
		type: ElementCellLayout
		_delete (func ptr): 0xb7f331b3
		_print (func ptr): 0xb7f331dc
		_copy (func ptr): 0xb7f33250
	name: elementCellLayout
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7f32da0
	_construct (func ptr): 0xb7f33328
	_build (func ptr): 0xb7f333a5
	_initialise (func ptr): 0xb7f3345a
	_execute (func ptr): 0xb7f33498
	_destroy (func ptr): 0xb7f3349d
	isConstructed: True
	isBuilt: True
	isInitialised: True
	hasExecuted: False
	isDestroyed: False
	constructor function name: elementCellLayout-Construct
	build function name: elementCellLayout-Build
	initialise function name: elementCellLayout-Initialise
	execute function name: elementCellLayout-Execute
	destroy function name: elementCellLayout-Destroy
	_cellLocalCount (func ptr): 0xb7f334a2
	_cellShadowCount (func ptr): 0xb7f334e8
	_pointCount (func ptr): 0xb7f3352e
	_initialisePoints (func ptr): 0xb7f33594
	_mapElementIdToCellId (func ptr): 0xb7f3363d
	_isInCell (func ptr): 0xb7f33645
	_cellOf (func ptr): 0xb7f33699
	_getShadowInfo (func ptr): 0xb7f33711
	mesh (ptr): 0x8072c58
RandomParticleLayout (ptr): 0x807d658:
PerCellParticleLayout (ptr): 0x807d658
ParticleLayout (ptr): 0x807d658
Stg_Class (ptr): 0x807d658
	sizeOfSelf: 136
	_deleteSelf: Yes
	type: RandomParticleLayout
	_delete (func ptr): 0xb7f3d7e6
	_print (func ptr): 0xb7f3d80f
	_copy (func ptr): 0xb7f3d8a5
	_setInitialCounts (func ptr): 0xb7f3cbdd
	_initialiseParticles (func ptr): 0xb7f3cd92
	coordSystem: GlobalCoordSystem
	weightsInitialisedAtStartup : False
	_initialCount (func ptr): 0xb7f3da84
	_initialiseParticlesOfCell (func ptr): 0xb7f3da9b
	cellParticleCount: 4
	seed: 13
	cellLocalCount: 4
	cellDomainCount: 8
	cellShadowCount: 4
	cellPointCountTbl (ptr): 0x8070f98
	cellPointCountTbl[0-8]: 008 008 008 008 008 008 008 008 
	cellPointTbl (ptr): 0x807e9c8
	cellPointTbl[0-8]: {
	cellPointTbl[0][0-8]: {0 8 0} {150 8 0} {0 12 0} {150 12 0} {0 8 150} {150 8 150} {0 12 150} {150 12 150} 
	cellPointTbl[1][0-8]: {150 8 0} {300 8 0} {150 12 0} {300 12 0} {150 8 150} {300 8 150} {150 12 150} {300 12 150} 
	cellPointTbl[2][0-8]: {0 8 150} {150 8 150} {0 12 150} {150 12 150} {0 8 300} {150 8 300} {0 12 300} {150 12 300} 
	cellPointTbl[3][0-8]: {150 8 150} {300 8 150} {150 12 150} {300 12 150} {150 8 300} {300 8 300} {150 12 300} {300 12 300} 
	cellPointTbl[4][0-8]: {0 4 0} {150 4 0} {0 8 0} {150 8 0} {0 4 150} {150 4 150} {0 8 150} {150 8 150} 
	cellPointTbl[5][0-8]: {150 4 0} {300 4 0} {150 8 0} {300 8 0} {150 4 150} {300 4 150} {150 8 150} {300 8 150} 
	cellPointTbl[6][0-8]: {0 4 150} {150 4 150} {0 8 150} {150 8 150} {0 4 300} {150 4 300} {0 8 300} {150 8 300} 
	cellPointTbl[7][0-8]: {150 4 150} {300 4 150} {150 8 150} {300 8 150} {150 4 300} {300 4 300} {150 8 300} {300 8 300} 
}

	cellParticleCountTbl (ptr): 0x807eaf8
	cellParticleCountTbl[0-8]: 004 004 004 004 000 000 000 000 
	cellParticleSizeTbl (ptr): 0x807eb28
	cellParticleSizeTbl[0-8]: 004 004 004 004 000 000 000 000 
	cellParticleTbl (ptr): 0x807eb58
	cellParticleTbl [0-8]: {
		cellParticleTbl[0][0-4]:
			(part. index) 0
			(part. index) 1
			(part. index) 2
			(part. index) 3

		cellParticleTbl[1][0-4]:
			(part. index) 4
			(part. index) 5
			(part. index) 6
			(part. index) 7

		cellParticleTbl[2][0-4]:
			(part. index) 8
			(part. index) 9
			(part. index) 10
			(part. index) 11

		cellParticleTbl[3][0-4]:
			(part. index) 12
			(part. index) 13
			(part. index) 14
			(part. index) 15

		cellParticleTbl[4][0-0]:

		cellParticleTbl[5][0-0]:

		cellParticleTbl[6][0-0]:

		cellParticleTbl[7][0-0]:

	}
	particlesArraySize: 116
	particlesArrayDelta: 100
	extraParticlesFactor: 0.05

Printing coords of all local processor particles:
	local particle      0: ( 36.393,  10.685,  24.175)
	local particle      1: ( 106.19,  10.926,  13.188)
	local particle      2: ( 107.93,  8.6386,  64.247)
	local particle      3: ( 105.66,  11.401,  72.744)
	local particle      4: ( 261.92,  10.607,  15.099)
	local particle      5: ( 275.23,  8.3656,  23.032)
	local particle      6: ( 242.36,  10.759,  108.86)
	local particle      7: ( 201.31,  8.9555,  137.57)
	local particle      8: ( 91.249,  8.7051,  220.37)
	local particle      9: ( 40.359,  8.3716,  202.27)
	local particle     10: ( 60.611,   9.112,  204.88)
	local particle     11: ( 32.111,  11.509,  224.72)
	local particle     12: ( 196.11,  10.331,  252.69)
	local particle     13: ( 271.87,  9.6744,  211.88)
	local particle     14: ( 162.47,  10.951,  256.25)
	local particle     15: ( 218.88,   8.529,  272.07)

Printing coords of all local particles, cell-by-cell:
	Local Cell   0:
		part. InCell   0 (local index      0): ( 36.393,  10.685,  24.175)
		part. InCell   1 (local index      1): ( 106.19,  10.926,  13.188)
		part. InCell   2 (local index      2): ( 107.93,  8.6386,  64.247)
		part. InCell   3 (local index      3): ( 105.66,  11.401,  72.744)
	Local Cell   1:
		part. InCell   0 (local index      4): ( 261.92,  10.607,  15.099)
		part. InCell   1 (local index      5): ( 275.23,  8.3656,  23.032)
		part. InCell   2 (local index      6): ( 242.36,  10.759,  108.86)
		part. InCell   3 (local index      7): ( 201.31,  8.9555,  137.57)
	Local Cell   2:
		part. InCell   0 (local index      8): ( 91.249,  8.7051,  220.37)
		part. InCell   1 (local index      9): ( 40.359,  8.3716,  202.27)
		part. InCell   2 (local index     10): ( 60.611,   9.112,  204.88)
		part. InCell   3 (local index     11): ( 32.111,  11.509,  224.72)
	Local Cell   3:
		part. InCell   0 (local index     12): ( 196.11,  10.331,  252.69)
		part. InCell   1 (local index     13): ( 271.87,  9.6744,  211.88)
		part. InCell   2 (local index     14): ( 162.47,  10.951,  256.25)
		part. InCell   3 (local index     15): ( 218.88,   8.529,  272.07)
